src/v/datalake/translation/scheduling.cc

```diff
@@ -78,22 +79,95 @@ class default_reservations_tracker : public reservations_tracker {
+     * disk manager, and then waiting for sufficient disk resources to become
+     * available.
      */
     ss::future<reservation>
     reserve_disk(size_t, ss::abort_source& as) override {
+        const auto init_backoff = 100ms;
+        const auto max_backoff = 1000ms;
         const auto nr = _reservation_block_size;
-        auto opt_units = ss::try_get_units(_available_disk, nr);
-        if (opt_units) {
-            co_return std::move(opt_units.value());
-        }
-        _notifier.notify_disk_exhausted();
-        try {
-            co_return co_await ss::get_units(_available_disk, nr, as);
-        } catch (...) {
-            as.check();
-            std::rethrow_exception(std::current_exception());
+
+        auto backoff = init_backoff;
+        while (true) {
+            /*
+             * first we attempt to acquire disk reservation from what is
+             * available local on this shard. when a translator on this core
+             * finishes it will return its units to this pool. we use up all of
+             * the available units, so check that get_units returns a positive
+             * value because when requesting 0 units it will be granted but the
+             * system won't make progress.
+             */
+            auto opt_units = ss::try_get_units(
+              _available_disk, std::min(nr, _available_disk.current()));
+            if (opt_units.has_value() && opt_units.value().count() > 0) {
+                vlog(
+                  datalake_log.trace,
+                  "Allocated {} disk reservation from shard-local pool. Total "
+                  "available {}",
+                  human::bytes(opt_units.value().count()),
+                  human::bytes(_available_disk.current()));
+                co_return std::move(opt_units.value());
+            }
+
+            /*
+             * if we are unable to acquire units locally, then pop on over
+             * to core-0 and ask the disk manager if there are available
+             * units that we can have. deposit any units locally and try
+             * again. note that below we add a very short, intentional delay
+             * to prevent accidentally spinning. we avoid immediately taking the
+             * units and returning to avoid the possibility that the translators
+             * on a core continually take units from each other rather than
+             * giving the system an opportunity to redistribute the units. this
+             * concern, however, is speculative.
+             */
+            auto units = co_await _disk_manager.reserve();
+            if (units > 0) {
+                _available_disk.signal(units);
+                vlog(
+                  datalake_log.debug,
+                  "Received {} disk reservation from global pool. Total "
+                  "available {}",
+                  human::bytes(units),
+                  human::bytes(_available_disk.current()));
```

Comment:
is this missing a continue statement? if reservation succeeds why would we sleep below before handing off reservation to the caller?

Reply:
Good spot, but it is intentional (though i don't feel strongly about the decision). The logic behind it is explained in the comment above that code: https://github.com/redpanda-data/redpanda/pull/25766/files#diff-c1b95efc3de022a0b4d6f59136bc6ceaa3dc32ea5d27903aca5abdc874cf1912R118-R123. It's not a clear cut choice, but rather a conservative heuristic.

Reply:
It is still not clear to me why the sleep is necessary after we got reservations from the global pool. Shouldn't we sleep only when such reservation is not satisfied?

Reply:
Sorry! The line numbers in the link above were to the wrong comment (or the line numbers changed, i dunno). either way:

>not clear to me why the sleep is necessary after we got reservations from the global pool

It's not necessary, you're right we could return immediately. rather, it's conservative approach that we can definitely change if we want. The motivation is that if a translator is running at a high throughput then there is the possibility that it starves out every other translator on all cores. So we make the translator pause for some number of milliseconds after exhausting its current reservation (~ single digit MBs). The exactly parameters to the policy are up for debate, but I think it makes sense to have some strategy for trying to avoid starvation.

Reply:
Oh I missed this discussion, I put the same comment elsewhere

> The motivation is that if a translator is running at a high throughput then there is the possibility that it starves out every other translator on all cores. So we make the translator pause for some number of milliseconds after exhausting its current reservation 

this condition is hit even when the translator(s) are running at normal throughputs, right? For example, the very first reservation of bytes (first time a data is written to the parquet file) hits a backoff of 100ms, that seems unnecessary? Agree in principle that a backoff is needed if there are many reservations back to back.

Reply:
This seems like an easy point of confusion, and I get that there's a comment above explaining, but I think it's much less error prone to move the salient point of that comment to where the fallthrough is happening.

Also, I don't feel strongly about it yet, but maybe it would be less surprising to move the sleep into the `translator_disk_tracker`, and instead make the interface of the `reservations_tracker` also return a hint of a duration to sleep for, if the reservation tracker went to core 0 or somesuch.